### ISSUE TITLE
RPC: GLM-5.2 crashes on multi-node CUDA RPC - invalid data ptr / graph_compute failed

### ISSUE THREAD

**Issue by Cephei-OpenSource** [ggml-org/llama.cpp#26583]
## Bug Description

GLM-5.2 crashes deterministically when attempting distributed inference across multiple DGX Spark nodes via ggml-rpc. The crash occurs during graph computation with `invalid data ptr` on the worker side and `ggml_backend_rpc_buffer_get_tensor` abort on the orchestrator side. This appears analogous to #26424 (DeepSeek V4, Metal) but affects CUDA (sm_121).

## Environment

- **Hardware:** 4x NVIDIA DGX Spark (GB10, Grace Blackwell, 128GB unified memory each)
- **OS:** DGX OS (Ubuntu-based, aarch64)
- **CUDA:** sm_121 native build
- **llama.cpp:** master `5788b510a` (ggml 0.18.1), built with `-DGGML_CUDA=ON -DGGML_RPC=ON`
- **Network:** 1Gbit Ethernet between nodes
- **Model:** GLM-5.2 UD-IQ4_NL (also reproduced with UD-Q4_K_M), unsloth GGUF split files

## Setup

**Workers (3 nodes):**

```
./build/bin/ggml-rpc-server -c --host 192.168.1.XXX --port 50052
```

**Orchestrator (1 node):**

```
./build/bin/llama-server \
  -m ~/models/gguf/glm-5.2/UD-IQ4_NL/GLM-5.2-UD-IQ4_NL-00001-of-00009.gguf \
  --host 0.0.0.0 --port 8080 \
  -ngl 999 -c 4096 \
  --tensor-split 1,1,1,1 \
  --rpc 192.168.1.121:50052,192.168.1.111:50052,192.168.1.128:50052
```

## Expected Behavior

Model loads across 4 nodes and serves inference (works with Gemma-4-26B, Qwen3.6-35B, Mistral-Small-3.2 on identical setup).

## Actual Behavior

Weights distribute successfully (~113GB cache per worker), but crashes during graph_compute:

**Worker output:**

[create_node] invalid data ptr[graph_compute] failed to create graph node 30 (id=197768091778640)
Client connection closed

**Orchestrator output:**

ggml-rpc.cpp:519: Remote RPC server crashed or returned malformed response
recv failed (bytes_recv=0, size_to_recv=8)

Stack trace: `ggml_backend_rpc_buffer_get_tensor` → `ggml_backend_tensor_copy` → `ggml_backend_sched_graph_compute_async` → `llama_context::graph_compute` → `llama_context::decode` (during `common_init_from_params`).

## Reproducibility

- 100% deterministic — same worker, same graph node (30)
- Reproduced with UD-IQ4_NL (~348GB) and UD-Q4_K_M (~434GB)
- Reproduced with `-c 4096` and `-c 131072`
- Reproduced with and without `--no-warmup`
- All 4 nodes on identical commit (`5788b510a`, fresh pull)

## What Works

- Single-node llama.cpp on same hardware (Gemma, Qwen, Mistral)
- 2-node llama.cpp RPC with smaller models (Gemma-4-26B, ~51.8 tok/s)
- GLM-5.2 on single-node (4×RTX 6000, Ollama, Note: Different machine! But same model GLM 5.2 UD-IQ4_NL)

## Related

- #26424 — DeepSeek V4 Flash RPC crash (Metal, similar MoE/MLA architecture)
- GLM-5.2 uses IndexShare and MTP/NextN layers — likely unsupported ops in RPC graph distribution

## Logs

Full orchestrator and worker terminal output, dmesg are attached.

[glm-rpc-orch.log](https://github.com/user-attachments/files/30708516/glm-rpc-orch.log)

[glm-rpc-worker3.log](https://github.com/user-attachments/files/30708521/glm-rpc-worker3.log)

[glm-rpc-dmesg-g3.log](https://github.com/user-attachments/files/30708525/glm-rpc-dmesg-g3.log)

**Comment by alexsde** [comment]
I'm seeing the same on multi-node Strix Halo setup (3x AMD 395+ with 128GB of unified memory each) on the latest commit. Deepseek v4 flash (both Q4_K_XL and Q8_K_XL) and GLM 5.2 (Q2_K_XL) crash with the same error, which also crashes the worker entirely, not just closes the connection. At the same time Qwen 3.5 397B (Q4_K_XL) works fine.

**Comment by ludwigjossieaux** [comment]
I can confirm it with a different heterogeneous 3-node cluster setup : EPYC head with 4 x RTX PRO 6000 Blackwell (SM120) + 2 x ASUS Ascent GX10 (GB10, SM121) as RPC workers, connected point-to-point with 200G DAC over ConnectX-7. ROCEv2 RDMA is confirmed active in the worker logs.

Short version: Running GLM-5.2 over multi-node RPC dies exactly as described here - "[create_node] invalid data ptr[graph_compute] failed to create graph node 30 (id=98582056112224)" on a worker, head aborting in ggml_backend_tensor_copy -> ggml_backend_sched_graph_compute_async -> llama_decode. This is 100% deterministic at each run.

For the long version, because I elimited some suspects along the way:
- Not the network/fabric : Qwen3.5-397B and Qwen3.5-122B run perfectly over the same RPC setup, RDMA and all.
- Not corruption : Crashed on two independently downloaded quants (unsloth UD-Q4_K_XL and UD_IQ4_XS)
- Not the split strategy : crashed with explicit --tensor-split 19,19,19,19,12,12, with boundaries aligned to 4-layer IndexShare groups (20.5,20.5,20.5,20.5,10.3,7.7), and with no split flags at all (fit auto). As it was mentionned in the related of section the message, I really wanted IndexShare to be the problem, but looks like it's not
- Not llama-server specific : llama-cli crashed at the same node 30
- Not MTP/NextN loading : in every build i tested, blk.78 is logged as unused/ignored

The bisect : first I suspected the b10182 GLM MTP wiring and went back to b10719, but it still crashed. So I needed a good known endpoint. The NVDIA DGX Spark forum has people running GLM-5.2 UD-IQ4_XS over 4-node GB10 RPC around June 20-25 (fresh git clone of master at that time) (link : https://forums.developer.nvidia.com/t/glm-5-2-iq4-xs-on-4x-gb10-6-28-tok-s-dsa-active-full-recipe/373933/6) , and the release history puts b9767 on June 22 - right in the window. So i checked out b9767 and it works : same model files, same cluster, loads in minutes, serves coherent output at 14tk/s decode. Rolled forward to master to be sure -> node 30. Rolled back -> works again.

One caveat : b9767 is the nearest tag of "master as of about June 20-22" based on the forum post, not a proven exact boundary, so it looks like the breaking commit would be inside that window.

Regression window i could identify : b9767 (good) -> b10173 (bad), about 470 commits. Reproduces in minutes on my setup once caches are warm.

Happy to run the actual bisect in that window or test any patch

**Comment by Error00101** [comment]
I have 2 Strix Halo nodes running (Trying to run Deepseek v4 flash 0731 - Q8_K_XL) same crash.
Tried running a few other models: They work.
I tried the first commit to support Deepseek v4 (Jun 29 - sha: 8c146a8366304c871efc26057cc90370ccf58dad / b9840) to run it did not work either.
As far as i can tell: Deepseek never worked over RPC.

**Comment by lich-77** [comment]
I have tested that https://github.com/ggml-org/llama.cpp/pull/26500, resolves this crash when deploying DeepSeek on 3RPC nodes.

**Comment by ludwigjossieaux** [comment]
I tested too #26500 patched on master on the full setup (EPYC head, 4x RTX PRO 6000 SM120 + 2x GB10 SM121 workers, RoCEv2 RDMA) and GLM-5.2  UD-IQ4_XS now loads and serves correctly over multi-node RPC. no more graph node 30 / invalid data ptr.

Bonus: MTP speculative decoding (--spec-type draft-mtp --spec-draft-n-max 2) alors works over RPC. No other optimization yet, but got 14tk/s -> 20tk/s decode, draft acceptance 0.63. Tested at 256k context.

**Comment by Cephei-OpenSource** [comment]
Confirming that [26500](https://github.com/ggml-org/llama.cpp/pull/26500) (f0c483c applied on current master, same build on all nodes) fixes this crash for GLM-5.2 on 4x DGX Spark (GB10, sm_121) — 1 orchestrator + 3 RPC workers.

Before: deterministic crash during warmup, exactly as in the original report. After: the model loads, warmup passes, and generation runs stable over long outputs. Thanks @hmirin!

**Comment by alexsde** [comment]
Can confirm that DeepSeek v4 0731 (unsloth/DeepSeek-V4-Flash-0731-UD-Q8_K_XL) does run on 3x Strix Halo RPC nodes with the fix from the above commit applied, however it generates complete garbage. 
GLM 5.2 (unsloth/GLM-5.2-UD-Q2_K_XL) works completely fine now though.

Log from a query to DeepSeek model:
```
15.30.258.110 I srv  server_strea: conv_id= (empty=1)
15.30.258.407 I srv    operator(): chat format: peg-native
15.30.258.643 I slot get_availabl: id  0 | task -1 |  - checking sim = 1.000 (11/11) > 0.100
15.30.258.647 I slot get_availabl: id  0 | task -1 | selected slot by LCP similarity, f_sim_best = 1.000 (> 0.100 thold), f_keep = 0.289
15.30.258.648 I srv  get_availabl: updating prompt cache
15.30.258.721 I srv   prompt_save:  - saving prompt with length 38, total state size = 13.474 MiB (draft: 0.000 MiB)
15.30.427.413 I srv          load:  - looking for better prompt, base f_keep = 0.289, f_sim = 1.000
15.30.427.418 I srv          load:    - prompt with length      38, lcp =      11, f_keep = 0.289, f_sim = 1.000
15.30.427.423 I srv        update:  - cache state: 1 prompts, 25.413 MiB (limits: 32000.000 MiB, 32000 tokens, 47849 est)
15.30.427.424 I srv        update:    - prompt 0x33c69030:      38 tokens, checkpoints:  1,    25.413 MiB
15.30.427.425 I srv  get_availabl: prompt cache update took 168.78 ms
15.30.427.510 I slot launch_slot_: id  0 | task -1 | sampler chain: logits -> ?penalties -> ?dry -> ?top-n-sigma -> top-k -> ?typical -> ?top-p -> ?min-p -> ?xtc -> ?temp-ext -> dist 
15.30.427.517 I slot launch_slot_: id  0 | task -1 | sampler params: 
	repeat_last_n = 64, repeat_penalty = 1.000, frequency_penalty = 0.000, presence_penalty = 0.000
	dry_multiplier = 0.000, dry_base = 1.750, dry_allowed_length = 2, dry_penalty_last_n = 64
	top_k = 40, top_p = 1.000, min_p = 0.000, xtc_probability = 0.000, xtc_threshold = 0.100, typical_p = 1.000, top_n_sigma = -1.000, temp = 1.000
	mirostat = 0, mirostat_lr = 0.100, mirostat_ent = 5.000, adaptive_target = -1.000, adaptive_decay = 0.900
15.30.427.519 I slot launch_slot_: id  0 | task 31 | processing task, is_child = 0
15.30.427.532 I slot   operator(): id  0 | task 31 | new prompt, n_ctx_slot = 32000, n_keep = 0, task.n_tokens = 11
15.30.427.537 I slot   operator(): id  0 | task 31 | checking checkpoint with [6, 6] against 10...
15.30.432.508 I slot   operator(): id  0 | task 31 | restored context checkpoint (pos_min = 6, pos_max = 6, n_tokens = 7, n_past = 7, size = 11.939 MiB)
15.30.432.515 I slot   operator(): id  0 | task 31 | cached n_tokens = 7, memory_seq_rm [7, end)
15.30.432.568 I slot init_sampler: id  0 | task 31 | init sampler, took 0.00 ms, tokens: text = 11, total = 11
15.30.577.206 I slot create_check: id  0 | task 31 | created context checkpoint 2 of 32 (pos_min = 6, pos_max = 6, n_tokens = 7, size = 11.939 MiB)
15.41.614.867 I slot print_timing: id  0 | task 31 | n_gen =    100, tg =   9.25 t/s, tg_3s =   9.34 t/s
15.44.657.751 I slot print_timing: id  0 | task 31 | n_gen =    125, tg =   9.02 t/s, tg_3s =   8.22 t/s
15.45.525.168 W srv          stop: cancel task, id_task = 31
15.45.632.102 I slot      release: id  0 | task 31 | stop processing: n_tokens = 144, truncated = 0
15.45.632.123 I srv  update_slots: all slots are idle
```

<img width="1123" height="431" alt="Image" src="https://github.com/user-attachments/assets/88d9fe36-1f2d-4bb0-920c-8663a53e125b" />

**Comment by alexsde** [comment]
An update: got around to testing GLM 5.2 and while it does load like I mentioned in the last comment, it crashes whenever I try to send any message, no matter how I split the tensors across the RPC nodes. I even tried offloading majority of the model onto the two non-host nodes in case it was an issue with cache size, but even with an almost empty memory on the host it still crashes. And to be more specific - host node crashes, not the ones running on other machines.
I know for the fact that this model runs on my setup because I was able to use it in the past - it broke around the same time as DeepSeek issue started happening.

```
ggml_vulkan: Device memory allocation of size 8589934592 failed.
ggml_vulkan: Requested buffer size exceeds device buffer size limit: ErrorOutOfDeviceMemory
[New LWP 1561717]
[New LWP 1561716]

This GDB supports auto-downloading debuginfo from the following URLs:
  <ima:enforcing>
  <https://debuginfod.fedoraproject.org/>
  <ima:ignore>
Enable debuginfod for this session? (y or [n]) [answered N; input not from terminal]
Debuginfod has been disabled.
To make this setting permanent, add 'set debuginfod enabled off' to .gdbinit.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
0x00007f8a2dd6b312 in __syscall_cancel_arch () from /lib64/libc.so.6
#0  0x00007f8a2dd6b312 in __syscall_cancel_arch () from /lib64/libc.so.6
#1  0x00007f8a2dd5f52c in __internal_syscall_cancel () from /lib64/libc.so.6
#2  0x00007f8a2dd5f574 in __syscall_cancel () from /lib64/libc.so.6
#3  0x00007f8a2ddcf14f in wait4 () from /lib64/libc.so.6
#4  0x00007f8a2e33205b in ggml_print_backtrace () from /home/shin/llama.cpp/build/bin/libggml-base.so.0
#5  0x00007f8a2e3457c9 in ggml_uncaught_exception() () from /home/shin/llama.cpp/build/bin/libggml-base.so.0
#6  0x00007f8a2e020b1c in __cxxabiv1::__terminate(void (*)()) () from /lib64/libstdc++.so.6
#7  0x00007f8a2e009159 in std::terminate() () from /lib64/libstdc++.so.6
#8  0x00007f8a2e020dc8 in __cxa_throw () from /lib64/libstdc++.so.6
#9  0x00007f8a2e40f414 in ggml_vk_create_buffer_device(std::shared_ptr<vk_device_struct>&, unsigned long) [clone .cold] () from /home/shin/llama.cpp/build/bin/libggml-vulkan.so.0
#10 0x00007f8a2e64037a in ggml_vk_preallocate_buffers(ggml_backend_vk_context*, std::shared_ptr<vk_context_struct>) () from /home/shin/llama.cpp/build/bin/libggml-vulkan.so.0
#11 0x00007f8a2e641ded in ggml_vk_topk(ggml_backend_vk_context*, std::shared_ptr<vk_context_struct>&, ggml_tensor const*, ggml_tensor*) () from /home/shin/llama.cpp/build/bin/libggml-vulkan.so.0
#12 0x00007f8a2e681b6b in ggml_vk_build_graph(ggml_backend_vk_context*, ggml_cgraph*, int, ggml_tensor*, int, bool, bool, bool) [clone .isra.0] () from /home/shin/llama.cpp/build/bin/libggml-vulkan.so.0
#13 0x00007f8a2e6828a4 in ggml_backend_vk_graph_compute(ggml_backend*, ggml_cgraph*) () from /home/shin/llama.cpp/build/bin/libggml-vulkan.so.0
#14 0x00007f8a2e34a62e in ggml_backend_graph_compute () from /home/shin/llama.cpp/build/bin/libggml-base.so.0
#15 0x00007f8a3178971d in rpc_server::graph_compute(std::vector<unsigned char, std::allocator<unsigned char> > const&) () from /home/shin/llama.cpp/build/bin/libggml-rpc.so.0
#16 0x00007f8a31792747 in rpc_serve_client(std::vector<ggml_backend*, std::allocator<ggml_backend*> > const&, char const*, std::shared_ptr<socket_t>) () from /home/shin/llama.cpp/build/bin/libggml-rpc.so.0
#17 0x00007f8a3179572c in ggml_backend_rpc_start_server () from /home/shin/llama.cpp/build/bin/libggml-rpc.so.0
#18 0x000000000040231c in main ()
[Inferior 1 (process 1561715) detached]
terminate called after throwing an instance of 'vk::SystemError'
  what():  Requested buffer size exceeds device buffer size limit: ErrorOutOfDeviceMemory
Aborted                    (core dumped) build/bin/ggml-rpc-server -c --port 9000
```
